### PR TITLE
[MAINTENANCE] Cleanup/refactor prerequisite for group/filter/sort Expectations by domain

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* A small refactor for a better code readability.
* Correcting the wording in certain warning messages.
* Fixing the directive for the `quantiles` metric.
* Making certain `DataAssistant` fields private (which should not be public properties).
* Adding type hints in a few places, where they were missing.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-761/GREAT-1082
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by ghost** [comment]
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5523&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5523&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=5523&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map Legend

File: great_expectations/core/expectation_configuration.py [hunk 1]
```diff
@@ -1328,9 +1338,7 @@ def get_evaluation_parameter_dependencies(self) -> dict:
             try:
                 urn = ge_urn.parseString(string_urn)
             except ParserError:
-                logger.warning(
-                    f"Unable to parse great_expectations urn {value['$PARAMETER']}"
-                )
+                logger.warning("Unable to parse great_expectations urn['$PARAMETER']")
```
Comment (NathanFarmer) [inline great_expectations/core/expectation_configuration.py:1341]: What is wrong with providing this urn value?

File: great_expectations/core/expectation_suite.py [hunk 2]
```diff
@@ -667,17 +667,44 @@ def add_expectation(
...
+        """Return a list of column_pair map expectations."""
+        return [
+            e
+            for e in self.expectations
+            if "column_A" in e.kwargs and "column_B" in e.kwargs
+        ]
+
+    def get_multicolumn_expectations(self) -> List[ExpectationConfiguration]:
+        """Return a list of multicolumn map expectations."""
+        return [e for e in self.expectations if "column_list" in e.kwargs]
```
Comment (NathanFarmer) [inline great_expectations/core/expectation_suite.py:692]: ❤️

**Review by NathanFarmer — COMMENTED** [review]
Why did we drop the URN value from the f-string in the error message?

File: great_expectations/core/expectation_configuration.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/core/expectation_configuration.py:1341]: @NathanFarmer It was a bug (use before define!).

File: great_expectations/core/expectation_configuration.py [hunk 3]
```diff
@@ -1373,7 +1382,7 @@ def validate(
         validator: Any,  # Can't type as Validator due to import cycle
         runtime_configuration=None,
     ):
-        expectation_impl = self._get_expectation_impl()
+        expectation_impl: "Expectation" = self._get_expectation_impl()  # noqa: F821
```
Comment (Kilo59) [inline great_expectations/core/expectation_configuration.py:1385]: `undefined name name`
https://flake8.pycqa.org/en/latest/user/error-codes.html

Do we just need to import `annotations` from `__future__` to address this?

```
from __future__ import annotations
```

As an aside, one thing I do dislike about flake8 compared to pylint is the obscure error codes and related error suppression.
```
some_thing # noqa: F1234
```
vs
```
some_thing # noqa: human-readable-error-name
```

**Review by Kilo59 — APPROVED** [review]
LGTM.

I think we might be able to address the flake8 errors. But if not, NBD.

**Review by NathanFarmer — APPROVED** [review]
LGTM!

File: great_expectations/core/expectation_configuration.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/core/expectation_configuration.py:1385]: @Kilo59 Unfortunately, this is circular imports, and it is getting more difficult to manage them.

**Pull request merged**